From 76ce61d4ce41361d4f152545af644aea89133550 Mon Sep 17 00:00:00 2001
From: Shawn Willden <swillden@google.com>
Date: Fri, 27 Mar 2020 11:52:01 -0600
Subject: [PATCH 06/15] Fixed encoding of device_locked field

The attestation code used boringssl's ASN.1 encoding tools
incorrectly, causing it to encode incorrect values in device_locked.

Bug: b/152503089
Test: atest com.android.cts.devicepolicy.MixedDeviceOwnerTest#testKeyManagement
Merged-In: Ibe78d65719f78a4d64581bdef6988f1dfbd85523
Change-Id: Ibe78d65719f78a4d64581bdef6988f1dfbd85523
---
 include/keymaster/attestation_record.h | 2 +-
 km_openssl/attestation_record.cpp      | 3 +--
 2 files changed, 2 insertions(+), 3 deletions(-)

diff --git a/include/keymaster/attestation_record.h b/include/keymaster/attestation_record.h
index 2a6f24b..17a6429 100644
--- a/include/keymaster/attestation_record.h
+++ b/include/keymaster/attestation_record.h
@@ -43,7 +43,7 @@ struct ASN1_TYPE_Delete {
 
 typedef struct km_root_of_trust {
     ASN1_OCTET_STRING* verified_boot_key;
-    ASN1_BOOLEAN* device_locked;
+    ASN1_BOOLEAN device_locked;
     ASN1_ENUMERATED* verified_boot_state;
     ASN1_OCTET_STRING* verified_boot_hash;
 } KM_ROOT_OF_TRUST;
diff --git a/km_openssl/attestation_record.cpp b/km_openssl/attestation_record.cpp
index d1a920b..4f70138 100644
--- a/km_openssl/attestation_record.cpp
+++ b/km_openssl/attestation_record.cpp
@@ -417,8 +417,7 @@ build_attestation_record(const AuthorizationSet& attestation_params, Authorizati
         return TranslateLastOpenSslError();
     }
 
-    root_of_trust->device_locked = reinterpret_cast<int*>(malloc(sizeof(ASN1_BOOLEAN)));
-    *root_of_trust->device_locked = device_locked;
+    root_of_trust->device_locked = device_locked ? 0xFF : 0x00;
     if (!ASN1_ENUMERATED_set(root_of_trust->verified_boot_state, verified_boot_state)) {
         return TranslateLastOpenSslError();
     }
-- 
2.17.1

